Update cryptography to 41.0.0

zenmonkeykstop opened the pull request:
## Status

Ready for review 

## Description of Changes

Fixes #6856

Updates cryptography from 39.0.1 to 41.0.0 to clear alerts related to its openssl dependency.



## Testing

- [x] CI is passing
- [x] basic SI and JI functionality is working

## Checklist
Choose one of the following:

- [x] I have performed a diff review and pasted the contents to [the packaging wiki](https://github.com/freedomofpress/securedrop-builder/wiki/cryptography-39.0.1-to-41.0.0)
- [x] I would like someone else to do the diff review

zenmonkeykstop committed:
Updated cryptography to 41.0.0

legoktm commented:
Seems like the cryptography Rust code is linking with openssl now, so we need to install `pkg-config` in the builder:

```
    run pkg_config fail: Could not run `PKG_CONFIG_ALLOW_SYSTEM_CFLAGS="1" "pkg-config" "--libs" "--cflags" "openssl"`
    The pkg-config command could not be found.

    Most likely, you need to install a pkg-config package for your OS.
    Try `apt install pkg-config`, or `yum install pkg-config`,
    or `pkg install pkg-config`, or `apk add pkgconfig` depending on your distribution.

    If you've already installed it, ensure the pkg-config command is one of the
    directories in the PATH environment variable.

    If you did not expect this build to link to a pre-installed system library,
    then check documentation of the openssl-sys crate for an option to
    build the library from source, or disable features or dependencies
    that require pkg-config.

    --- stderr
    thread 'main' panicked at '

    Could not find directory of OpenSSL installation, and this `-sys` crate cannot
    proceed without this knowledge. If OpenSSL is installed and this crate had
    trouble finding it,  you can set the `OPENSSL_DIR` environment variable for the
    compilation process.

    Make sure you also have the development packages of openssl installed.
    For example, `libssl-dev` on Ubuntu or `openssl-devel` on Fedora.

    If you're in a situation where you think the directory *should* be found
    automatically, please open a bug at https://github.com/sfackler/rust-openssl
    and include information about your system as well as this message.

    $HOST = x86_64-unknown-linux-gnu
    $TARGET = x86_64-unknown-linux-gnu
    openssl-sys = 0.9.88


    It looks like you're compiling on Linux and also targeting Linux. Currently this
    requires the `pkg-config` utility to find OpenSSL but unfortunately `pkg-config`
    could not be found. If you have OpenSSL installed you can likely fix this by
    installing `pkg-config`.

    ', /opt/cargo/registry/src/index.crates.io-6f17d22bba15001f/openssl-sys-0.9.88/build/find_normal.rs:190:5
    note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
  warning: build failed, waiting for other jobs to finish...
```

zenmonkeykstop commented:
There's a failure from earlier against a branch without this cryptography bump, so I'd expect to have been seeing this for a while: https://app.circleci.com/pipelines/github/freedomofpress/securedrop/5913/workflows/5b64bd2a-a6af-4cf1-9628-e7cd29b4d45f/jobs/69563

legoktm commented:
That one is a different failure in the OSSEC build step, which I think I know the cause of...

legoktm commented:
From https://cryptography.io/en/latest/changelog/#v40-0-0
> BACKWARDS INCOMPATIBLE: As announced in the 39.0.0 changelog, the way cryptography links OpenSSL has changed. This only impacts users who build cryptography from source (i.e., not from a wheel), and specify their own version of OpenSSL. For those users, the CFLAGS, LDFLAGS, INCLUDE, LIB, and CRYPTOGRAPHY_SUPPRESS_LINK_FLAGS environment variables are no longer valid. Instead, users need to configure their builds [as documented here](https://docs.rs/openssl/latest/openssl/#automatic).

legoktm committed:
builder: Add pkg-config for Rust's openssl crate

Rust's openssl crate is now used by cryptography and soon by Sequoia.
To link with the system openssl, pkg-config needs to be installed at
build-time.

See <https://docs.rs/openssl/0.10.54/openssl/index.html#automatic>.

eaon commented:
Added diff-review to https://github.com/freedomofpress/securedrop-builder/wiki/cryptography-39.0.1-to-41.0.0 - have nothing of note to report that the `CHANGELOG` doesn't already cover.

legoktm commented on securedrop/requirements/python3/develop-requirements.in:
```diff
@@ -8,6 +8,7 @@ black
 boto
 boto3
 cffi>=1.14.2
+cryptography>=41.0.0
```
Why do we also need to add this here?

zenmonkeykstop replied:
> Why do we also need to add this here?

in the absence of the Ansible bump, it was still pulling in the older version.
